### PR TITLE
Drinks can now carry diseases, Wipe down your glasses with rags bartenders!

### PR DESCRIPTION
**Pull request by Couls**
## About The Pull Request
Drinks can now carry diseases that spread through fluids

## Why It's Good For The Game
Gives the damp rag more uses, also makes sense that saliva would spread viruses

## Changelog
:cl:
add: Recent reports have seen increased rates of infection from reused glasses. Cleaning them with soap or rags drastically decreases chance of infection.
/:cl:

### CODE REVIEW

**Comment by OmegaDarkPotato** [comment]
Honestly makes sense but I'm 50/50 on this
On the one hand, it makes a reason for me to wipe down the glasses I hand out aside from just doing it because it's a slow shift, but on the other hand, it makes it so I *have* to wipe do--

Wait a minute I could totally make glasses spread lethal diseases as a virologist now I love this PR

**Comment by ATH1909** [comment]
I wonder if we could do a similar thing with syringes and blood+ transmission viruses (so, basically all custom viruses) at some point as well.

**Comment by TemporalOroboros** [comment]
Oh god, contaminated omnizine hypospray sounds like an incredibly sneaky way to distribute a plague.

**Comment by Doosteroni** [comment]
Don’t re-use needles kiddies

**Comment by interestingusernam3** [comment]
Make it work on beakers too please. I want to be able to exterminate assistants with contaminated beakers of meth.

**Comment by Couls** [comment]
> Make it work on beakers too please. I want to be able to exterminate assistants with contaminated beakers of meth.

Already does, but someone infected has to drink from it first

**Comment by necromanceranne** [comment]
Typhoid Mary update.

**Comment by ArcaneMusic** [comment]
It would be based if you extended dish-cleaning to the dish-drive so it has more of a use.
Either way, very cool.

**Comment by 1glitchycent** [comment]
This seems to be a pretty useless feature tbh, it doesn't add much to the game besides tedium, really.

**Comment by KathyRyals** [comment]
> Oh god, contaminated omnizine hypospray sounds like an incredibly sneaky way to distribute a plague.

Sounds cool but the hypospray doesn't use a needle so it shouldn't spread disease

**Comment by TemporalOroboros** [comment]
> Sounds cool but the hypospray doesn't use a needle so it shouldn't spread disease

https://en.wikipedia.org/wiki/Jet_injector#Splash-back

> Numerous studies have found cross-infection of diseases from jet injections. An experiment using mice, published in 1985, showed that jet injectors would frequently transmit the viral infection lactate dehydrogenase elevating virus (LDV) from one mouse to another. Another study used the device on a calf, then tested the fluid remaining in the injector for blood. Every injector they tested had detectable blood in a quantity sufficient to pass on a virus such as hepatitis B.

> From 1984 to 1985, a weight-loss clinic in Los Angeles administered human chorionic gonadotropin (hCG) with a Med-E-Jet injector. CDC investigation found 57 out of 239 people who had received the jet injection tested positive for hepatitis B.

> Jet injectors have also been found to inoculate bacteria from the environment into users. In 1988 a podiatry clinic used a jet injector to deliver local anaesthetic into patients' toes. Eight of these patients developed infections caused by Mycobacterium chelonae. The injector was stored in a container of water and disinfectant between use, but the organism grew in the container. This species of bacteria is sometimes found in tap water, and had been previously associated with infections from jet injectors.

**Comment by KathyRyals** [comment]
Okay but we are using an hyper advanced hypospray, one of a kind on the station. I would hope that a few centuries later, we'd have done away with this issue !

File: code/datums/components/infective.dm [hunk 1]
```diff
@@ -26,6 +25,10 @@
 		RegisterSignal(parent, COMSIG_ITEM_EQUIPPED, .proc/try_infect_equipped)
 		if(istype(parent, /obj/item/reagent_containers/food/snacks))
 			RegisterSignal(parent, COMSIG_FOOD_EATEN, .proc/try_infect_eat)
+		if(istype(parent, /obj/item/reagent_containers/food/drinks))
+			RegisterSignal(parent, COMSIG_DRINK_DRANK, .proc/try_infect_drink)
+		if(istype(parent, /obj/item/reagent_containers/glass))
```
Comment (Rohesie) [inline code/datums/components/infective.dm:30]:
```suggestion
		else if(istype(parent, /obj/item/reagent_containers/food/drinks))
			RegisterSignal(parent, COMSIG_DRINK_DRANK, .proc/try_infect_drink)
		else if(istype(parent, /obj/item/reagent_containers/glass))
```

File: code/datums/components/infective.dm [hunk 2]
```diff
@@ -34,6 +37,11 @@
 		eater.ForceContractDisease(V)
 	try_infect(feeder, BODY_ZONE_L_ARM)
 
+/datum/component/infective/proc/try_infect_drink(datum/source, mob/living/drinker, mob/living/feeder)
+	for(var/V in diseases)
+		drinker.ForceContractDisease(V)
```
Comment (Rohesie) [inline code/datums/components/infective.dm:42]:
```suggestion
	for(var/disease in diseases)
		drinker.ForceContractDisease(disease)
```

File: code/datums/components/infective.dm [hunk 3]
```diff
@@ -34,6 +37,11 @@
...
+	try_infect(feeder, BODY_ZONE_L_ARM)
```
Comment (Rohesie) [inline code/datums/components/infective.dm:43]: Why the left arm?

File: code/modules/food_and_drinks/drinks/drinks.dm [hunk 4]
```diff
@@ -44,12 +44,22 @@
...
 	reagents.trans_to(M, gulp_size, transfered_by = user)
 	playsound(M.loc,'sound/items/drink.ogg', rand(10,50), TRUE)
+	if(iscarbon(M))
+		var/mob/living/carbon/C = M
+		var/list/diseases = C.get_static_viruses()
+		if(LAZYLEN(diseases))
+			var/list/datum/disease/diseases_to_add = list()
+			for(var/datum/disease/D in diseases)
+				if(D.spread_flags & DISEASE_SPREAD_CONTACT_FLUIDS)
+					diseases_to_add += D
```
Comment (Rohesie) [inline code/modules/food_and_drinks/drinks/drinks.dm:60]: Try better variable names.
```suggestion
		var/mob/living/carbon/carbon_drinker = M
		var/list/diseases = carbon_drinker.get_static_viruses()
		if(LAZYLEN(diseases))
			var/list/datum/disease/diseases_to_add = list()
			for(var/d in diseases)
				var/datum/disease/malady
				if(malady.spread_flags & DISEASE_SPREAD_CONTACT_FLUIDS)
					diseases_to_add += malady
```

File: code/modules/reagents/reagent_containers/glass.dm [hunk 5]
```diff
@@ -48,8 +48,19 @@
...
 			addtimer(CALLBACK(reagents, /datum/reagents.proc/trans_to, M, 5, TRUE, TRUE, FALSE, user, FALSE, INGEST), 5)
 			playsound(M.loc,'sound/items/drink.ogg', rand(10,50), TRUE)
+			if(iscarbon(M))
+				var/mob/living/carbon/C = M
+				var/list/diseases = C.get_static_viruses()
+				if(LAZYLEN(diseases))
+					var/list/datum/disease/diseases_to_add = list()
+					for(var/datum/disease/D in diseases)
+						if(D.spread_flags & DISEASE_SPREAD_CONTACT_FLUIDS)
+							diseases_to_add += D
```
Comment (Rohesie) [inline code/modules/reagents/reagent_containers/glass.dm:61]:
```suggestion
				var/mob/living/carbon/carbon_drinker = M
				var/list/diseases = carbon_drinker.get_static_viruses()
				if(LAZYLEN(diseases))
					var/list/datum/disease/diseases_to_add = list()
					for(var/d in diseases)
						var/datum/disease/malady = d
						if(malady.spread_flags & DISEASE_SPREAD_CONTACT_FLUIDS)
							diseases_to_add += malady
```

File: code/datums/components/infective.dm (hunk 3 above)
Comment (Couls) [reply code/datums/components/infective.dm:43]: No specific reason, the rest of the file used the left arm while infecting too and I thought to keep it uniform

File: code/datums/components/infective.dm [hunk 6]
```diff
@@ -34,6 +37,11 @@
 		eater.ForceContractDisease(V)
 	try_infect(feeder, BODY_ZONE_L_ARM)
 
+/datum/component/infective/proc/try_infect_drink(datum/source, mob/living/drinker, mob/living/feeder)
+	for(var/disease in diseases)
+		drinker.ForceContractDisease(disease)
+	try_infect(feeder, BODY_ZONE_L_ARM)
```
Comment (Rohesie) [inline code/datums/components/infective.dm:43]:
```suggestion
	var/appendage_zone
	switch(held_items.Find(source))
		if(0)
			appendage_zone = pick(BODY_ZONE_L_ARM, BODY_ZONE_R_ARM)
		if(1)
			appendage_zone = BODY_ZONE_L_ARM
		if(2)
			appendage_zone = BODY_ZONE_R_ARM
	try_infect(feeder, appendage_zone)
```

Comment (Couls) [reply code/datums/components/infective.dm:43]: No this makes sense, it forces the drinker to contract every disease that the drink has regardless. the one thats feeding the drinker has a chance of getting infected

Comment (Rohesie) [reply code/datums/components/infective.dm:43]: Ah, you are right, then it's on the hands. The `switch` version should take care of which hand is the correct one.
If you want to support multiple hands you could also do something like this:
```suggestion
	var/appendage_zone = held_items.Find(source)
	appendage_zone = appendage_zone == 0 ? pick(BODY_ZONE_L_ARM, BODY_ZONE_R_ARM)  : appendage_zone % 2 ? BODY_ZONE_R_ARM : BODY_ZONE_L_ARM
	try_infect(feeder, appendage_zone)
```

File: code/datums/components/infective.dm [hunk 7]
```diff
@@ -34,6 +37,13 @@
 		eater.ForceContractDisease(V)
 	try_infect(feeder, BODY_ZONE_L_ARM)
 
+/datum/component/infective/proc/try_infect_drink(datum/source, mob/living/drinker, mob/living/feeder)
+	for(var/disease in diseases)
+		drinker.ForceContractDisease(disease)
+	var/appendage_zone = held_items.Find(source)
```
Comment (Rohesie) [inline code/datums/components/infective.dm:43]:
```suggestion
	var/appendage_zone = feeder.held_items.Find(source)
```

**Comment by zeskorion** [comment]
... dude. Why did you use the infective component? Adding infective means nearly any interaction with the drink is as infective as drinking it

**Pull request merged**